File: Algorithms/Sorting/heap_sort.c
#include <stdio.h>
#include<malloc.h>
/*Name - Jyoti Mishra (jstjyoti)
Email-jyotimishra29091997@gmail.com
Heap-Sort (Divide and Conquer) 
INTRODUCTION :  
a. Definition of a complete Binary Tree: A  Tree is said to be complete binary tree if all it's levels have maximum number of nodes(two for each parent) except possibly the last level and 
all the nodes in the last level appear as per left as possible. 
b. Heap: A Heap is a Complete Binary Tree which is implemented using an array(Linear Data Structure). 
Max Heap: A heap is called max heap if all the nodes having a value greater than every children of the node. 
This is used for Sorting in Ascending Order. 
Min Heap: a heap is called min heap if all the nodes having smaller value than every children of the node*/

/*This function(swap()) is for swapping two values having Time Complexity O(1) and space Complexity O(1) as only one extra pointer required to swap values*/
void swap(int *x,int *y)
{
	int t=*x;
	*x=*y;
	*y=t;
}
/*This function(heap()) is responsible for creation a max heap everytime when it is called .
Space Complexity of this is O(1) as only a few single variables are needed and they together represent a constant number of extra variables
Time Complexity of above funtion is O(Logn) as the task is divided into two equal halves T(n) depends on Height as we are taking one of left and right subtrees to adjust the tree structure
Height of balanced tree having n nodes=Logn
*/
void heap(int *arr, int n, int i)
{
    int largest = i; 
    int l = 2*i + 1;  
    int r = 2*i + 2;  
 	if (l < n && *(arr+l) >*(arr+largest))
        largest = l;
 	if (r < n && *(arr+r) > *(arr+largest))
        largest = r;
   if (largest != i)
    {
        swap((arr+i), (arr+largest));
        heap(arr, n, largest);
    }
}
/*This Function(heapSort()) is for Calling heap() functtion and causing sort, each time by removing last element after first and last elements are swapped.
Space complexity of this is O(1) as each time only some constant number of Variables are needed.
Time Coomplexity is O(nLogn) as it is calling heap() function(discussed complexity above the function) from an i loop having iteration n times(below i loop)
first i loop has iterartions lesser than n times so it is having complexity about less than O(nLogn)
Therefore, Time Complexity is O(nLogn)
*/
void heapSort(int *arr, int n)
{
    int i;
/*this loop is making max-heap because it is starting from last 2nd level(level just above leaf nodes)as parent and arranging it into
heap and repeating the same upto the root node i.e.,heap function makes the root element having greatest value so,each parent having 
higher values than their child nodes and this is condition of Max-Heap
Time Complexity for this is O((n/2)*Logn) as the loop runs n/2 times and each time heap() function is called. So,this has a complexity
less than O(nlogn)*/
    for (i = n/2 - 1; i >= 0; i--)
	    heap(arr, n, i);
    for (i=n-1; i>=0; i--)
    {
        swap((arr+0), (arr+i));
 		heap(arr, i, 0);//arranging remaining elements in heap O(logn)
	   //loop having Complexity O(n log n)
    }
}
 /*This Function is only for Displaying array*/
void display(int *arr, int n)
{
	int i;
    for (i=0; i<n; ++i)
        printf("%d\t",*(arr+i));
    printf("\n");
}
/*in Main we take input number of elements and unsorted array and call function to sort the array and display*/
int main()
{
    int *arr,i,n;
	printf("Enter number of elements : ");
	scanf("%d",&n);
	arr=(int *)malloc(n*(sizeof(int)));
	printf("Enter elements : ");
	for(i=0;i<n;i++)
		scanf("%d",arr+i);
	 heapSort(arr, n);
 	 printf("Sorted array is : ");
    display(arr, n);
	return 0;
}

File: Algorithms/Sorting/Java/heapsort.java
import java.util.Scanner;

public class heapsort 
{   public  static void heapify(int[] a, int n, int pi)  
    {  
        //pi is the current index
        int largest = pi; 

        //l is the index of left child of pi when assumed in a binary tree.     
        int l = 2*pi + 1;   

        //r is the index of right child of pi when assumed in a binary tree.
        int r = 2*pi + 2; 

        if (l< n && a[l] > a[largest])  
        largest = l;  
        
        if (r < n && a[r] > a[largest])  
        largest = r;  
        
        //If element at pi is not the largest,swaping the elements.
        //Again calling heapify to place the largest correctly in the heap.
        if (largest != pi)  
        {  
            int t = a[pi];  
            a[pi]= a[largest];   
            a[largest] = t;  
            heapify(a, n, largest);  
        }  
    }  
      
    public static void sort(int[] arr, int len)  
    {    
        //Creating a max-heap by calling heapify.
        for (int i = len / 2 - 1; i >= 0; i--)  
        heapify(arr, len, i); 

        //Swaping the first element with the last element of the heap.
        //Thus largest element is placed at the last position.
        //(Heap till position i is taken into consideration at each step.)
        //Again calling heapify function to place the element at 0th position correctly in the heap.
        for (int i=len-1; i>=0; i--)  
        {  
            int temp = arr[0];  
            arr[0]= arr[i];   
            arr[i] = temp;  
            heapify(arr, i, 0);  
        }  
    }  
      
    public static void main(String []args)  
    {  
        Scanner sc=new Scanner(System.in);

        //Enter no. of elements
        int n=sc.nextInt();

        int a[]=new int[n];

        //Enter the elements
        for(int i=0;i<n;i++)
        a[i]=sc.nextInt();

        //Function to sort the elemntes using HeapSort.
        sort(a, 10);  
        
        //Printing the elements.
        for (int i=0; i<n; i++)  
        System.out.print(a[i]+" ");
    }  
}  

//Time Complexity : O(nLogn) => If heap is assumed as a binary tree,the height of the tree is (Log n).
//The heapify function works for at most (Log n) times.
//Heapify is called for all the elements of the array.Thus complexity is (nLogn).
//Space Complexity : O(n) => No extra space used.